From a04e211705dfbfbad4818587b5612ff5de436135 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@st.com>
Date: Mon, 9 May 2022 16:20:20 +0200
Subject: [PATCH 0256/1695] drivers: stm32_risaf: add RISAF faulty address
 dumping for debug

Add a debug function that dumps faulty addresses on all RISAF instances.

If there is a faulty Address in IADDRx register, the address is dumped.

Following the same logic, IAESRx is dumped if there's data in it.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git da2d5cff7078cfbe935c2762fdf513b21a416031 drivers: stm32_risaf: add RISAF faulty address dumping for debug]
Change-Id: Id5721b6449d5e90bc1c1a383f9ccc0767c2033a9
Signed-off-by: Gatien Chevallier <gatien.chevallier@st.com>
---
 core/arch/arm/plat-stm32mp2/platform_config.h |  1 +
 core/drivers/firewall/stm32_risaf.c           | 67 +++++++++++++++++++
 core/include/drivers/stm32_risaf.h            |  8 +++
 3 files changed, 76 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp2/platform_config.h b/core/arch/arm/plat-stm32mp2/platform_config.h
index 4d7af4a8c..e377d2ce0 100644
--- a/core/arch/arm/plat-stm32mp2/platform_config.h
+++ b/core/arch/arm/plat-stm32mp2/platform_config.h
@@ -54,6 +54,7 @@
 #define SAES_BASE			0x42050000
 #define PKA_BASE			0x42060000
 #define RIFSC_BASE			0x42080000
+#define RISAF4_BASE			0x420d0000
 #define RISAF5_BASE			0x420e0000
 #define RISAB6_BASE			0x42140000
 #define BSEC3_BASE			0x44000000
diff --git a/core/drivers/firewall/stm32_risaf.c b/core/drivers/firewall/stm32_risaf.c
index 2908b21d3..70e7e3fbe 100644
--- a/core/drivers/firewall/stm32_risaf.c
+++ b/core/drivers/firewall/stm32_risaf.c
@@ -19,6 +19,11 @@
 /* RISAF general registers (base relative) */
 #define _RISAF_CR			U(0x00)
 #define _RISAF_SR			U(0x04)
+#define _RISAF_IASR			U(0x08)
+#define _RISAF_IAESR0			U(0x20)
+#define _RISAF_IADDR0			U(0x24)
+#define _RISAF_IAESR1			U(0x28)
+#define _RISAF_IADDR1			U(0x2C)
 #define _RISAF_KEYR			U(0x30)
 #define _RISAF_HWCFGR			U(0xFF0)
 #define _RISAF_VERR			U(0xFF4)
@@ -121,6 +126,7 @@ struct stm32_risaf_pdata {
 	struct io_pa_va base;
 	struct clk *clock;
 	struct stm32_risaf_region *regions;
+	char risaf_name[20];
 	unsigned int nregions;
 	uintptr_t mem_base;
 	size_t mem_size;
@@ -172,6 +178,64 @@ static vaddr_t risaf_base(struct stm32_risaf_instance *risaf)
 	return io_pa_or_va_secure(&risaf->pdata.base, 1);
 }
 
+#ifdef CFG_TEE_CORE_DEBUG
+void stm32_risaf_dump_erroneous_data(void)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+
+	if (TRACE_LEVEL < TRACE_INFO)
+		return;
+
+	SLIST_FOREACH(risaf, &risaf_list, link) {
+		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
+
+		if (clk_enable(risaf->pdata.clock))
+			panic("Can't enable RISAF clock");
+
+		/* Check if faulty address on this RISAF */
+		if (!io_read32(base + _RISAF_IASR)) {
+			clk_disable(risaf->pdata.clock);
+			continue;
+		}
+
+		EMSG("\n\nDUMPING DATA FOR %s\n\n", risaf->pdata.risaf_name);
+		EMSG("=====================================================");
+		EMSG("Status register (IAESR0): %#"PRIx32,
+		     io_read32(base + _RISAF_IAESR0));
+
+		/* Reserved if dual port feature not available */
+		if (io_read32(base + _RISAF_IAESR1))
+			EMSG("Status register Dual Port (IAESR1) %#"PRIx32,
+			     io_read32(base + _RISAF_IAESR1));
+
+		EMSG("-----------------------------------------------------");
+		if (virt_to_phys((void *)base) == RISAF4_BASE) {
+			EMSG("Faulty address (IADDR0): %#"PRIxPA,
+			     risaf->pdata.mem_base +
+			     (paddr_t)io_read32(base + _RISAF_IADDR0));
+
+			/* Reserved if dual port feature not available */
+			if (io_read32(base + _RISAF_IADDR1))
+				EMSG("Dual port faulty address (IADDR1): %#"PRIxPA,
+				     risaf->pdata.mem_base +
+				     (paddr_t)io_read32(base + _RISAF_IADDR1));
+		} else {
+			EMSG("Faulty address (IADDR0): %#"PRIxPA,
+			     (paddr_t)io_read32(base + _RISAF_IADDR0));
+
+			/* Reserved if dual port feature not available */
+			if (io_read32(base + _RISAF_IADDR1))
+				EMSG("Dual port faulty address (IADDR1): %#"PRIxPA,
+				     (paddr_t)io_read32(base + _RISAF_IADDR1));
+		}
+
+		EMSG("=====================================================\n");
+
+		clk_disable(risaf->pdata.clock);
+	};
+}
+#endif
+
 static __maybe_unused
 bool risaf_is_hw_encryption_enabled(struct stm32_risaf_instance *risaf)
 {
@@ -412,6 +476,9 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 		return TEE_SUCCESS;
 	}
 
+	strncpy(risaf->pdata.risaf_name, fdt_get_name(fdt, node, NULL),
+		sizeof(risaf->pdata.risaf_name) - 1);
+
 	res = clk_enable(risaf->pdata.clock);
 	if (res)
 		goto err;
diff --git a/core/include/drivers/stm32_risaf.h b/core/include/drivers/stm32_risaf.h
index 7a39a1062..224004a35 100644
--- a/core/include/drivers/stm32_risaf.h
+++ b/core/include/drivers/stm32_risaf.h
@@ -11,4 +11,12 @@
 
 TEE_Result stm32_risaf_reconfigure(paddr_t base);
 
+#ifdef CFG_TEE_CORE_DEBUG
+void stm32_risaf_dump_erroneous_data(void);
+#else /* CFG_TEE_CORE_DEBUG */
+static inline void stm32_risaf_dump_erroneous_data(void)
+{
+}
+#endif /* CFG_TEE_CORE_DEBUG */
+
 #endif /*__DRIVERS_STM32_RISAF_H__*/
-- 
2.47.2

